tests/test_protocol.py

```diff
@@ -880,7 +880,7 @@ async def test_dict_with_optional_values(unused_tcp_port, postgres_db, database_
     """Test dict which may have None as a value"""
     configure(unused_tcp_port, database_name, postgres_db.port)
 
-    types = Union[int, str]
+    types = Union[pydantic.StrictInt, pydantic.StrictStr]
```

Comment:
This change is required because there is a conflict in the way that Pydantic and Python handle Union types:

* According to the Pydantic documentation, a value is evaluated against a union type by matching each subtype of the Union against the value. The type of the value in the Pydantic object is the first matching subtype. Example: If you match the String `"5"` against type `Union[int, str]` you will get an integer value while `Union[str, int]` results in a string value.
* The Python documentation on the other hand says that the order of the subtypes of a Union type is unimportant. As such, Python considers that `Union[str, int] == Union[int, str]`.

The problem with this test case is caused by the fact Python uses an internal type cache. This cache builds on top of the fact that the order of the subtypes of a Union is not important. This means that if the code mentions the type annotation `Optional[Union[str, int]]` somewhere, a second type annotation for `Optional[Union[int, str]]` (inverse order of subtypes), will result in a cache hit. Now it seems that `jinja2==3.0.1` does exactly that. It defines a type annotation `optional[Union[str, int]]` somewhere. This way the order of the subtypes in the model `Result` inverses to `Optional[Union[str, int]]` and the assertion `assert result.result["data"]["val"] == 5` fails, because the result evaluates to a string instead of an integer: `"5" != 5`.

Conclusion: Whenever a Union type is used in a Pydantic model, we should make sure that each subtype of the Union cannot coerce to any other subtype of that Union. Otherwise the type of the value in the Pydantic object is undefined.

A code snippet that isolates the issue for those who like:

```
from pydantic import BaseModel
from typing import Union, Optional

toggle = True  # Changing the value of this toggle result in different output

if toggle:
    # Populate the type cache
    Optional[Union[str, int]]


class Result(BaseModel):
    val: Optional[Union[int, str]]


r = Result(val=5)

print(type(r.val))
```

Reply:
wow!

Reply:
This issue was already reported to Pydantic via: https://github.com/samuelcolvin/pydantic/issues/1982